mono/metadata/threads.c

```diff
@@ -1634,7 +1634,8 @@ gint32 ves_icall_System_Threading_WaitHandle_WaitAll_internal(MonoArray *mono_ha
 
 	g_free(handles);
 
-	return ret;
+	/* WAIT_FAILED in waithandle.cs is different from WAIT_FAILED in Win32 API */
```

Comment:
The way it's done in CoreCLR is by never returning WAIT_FAILED, but in that case it would actually throw an exception directly from native (here: Thread::DoAppropriateWaitWorker)